docs/testing.rst

```diff
@@ -0,0 +1,78 @@
+Testing Services
+================
+
+Philosophy
+----------
+
+Nameko's conventions are designed to make testing as easy as possible. Services are likely to be small and single-purpose, and dependency injection makes it simple to replace and isolate pieces of functionality.
+
+Unit Testing
+------------
+
+Unit testing in nameko usually means testing a single service in isolation - i.e. without any or most of its dependencies.
+
+The :func:`~nameko.testing.services.worker_factory` utility will create a worker from a given service class, with its dependencies replaced by :class:`mock.MagicMock` objects. Dependency functionality can then be imitated by adding side effects and return values:
+
+.. literalinclude:: examples/testing/unit_test.py
+
+In some circumstances it's helpful to provide an alternative dependency, rather than use a mock. This may be a fully functioning replacement (e.g. a test database session) or a lightweight shim that provides partial functionality.
+
+.. literalinclude:: examples/testing/unit_test_alternative_dependency.py
+
+Integration Testing
+-------------------
+
+Integration testing in nameko means testing the interface between a number of services. The recommended way is to run all the services being tested in the normal way, and trigger behaviour by manually "firing" an entrypoint:
+
+.. literalinclude:: examples/testing/integration_test.py
+
+Interfaces that are out of scope for a particular test can be deactivated with one of the following test helpers:
+
+restrict_entrypoints
+^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: nameko.testing.services.restrict_entrypoints
+   :noindex:
+
+replace_injections
+^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: nameko.testing.services.replace_injections
+   :noindex:
+
+Complete Example
+^^^^^^^^^^^^^^^^
+
+The following integration testing example makes use of both scope-restricting helpers:
```

Comment:
Yeah, I will probably split this into different sections so it isn't so imposing.